fedimint/src/lib.rs

```diff
@@ -250,26 +284,52 @@ impl FedimintServer {
         // last signed epoch is at most 3 epochs before the next epoch + faulty nodes because
         // faulty nodes can withhold sigs for an epoch before getting banned
         let max_age: u64 = self.cfg.max_faulty() as u64 + 3;
+        let threshold = self.cfg.max_faulty() * 2 + 1;
+
+        // include our expected next_epoch as well in case we can contribute to the next consensus
+        let last_saved = self.consensus.db.get_value(&LastEpochKey);
+        let next_epoch = last_saved.expect("DB error").map(|e| e.0 + 1).unwrap_or(0);
+        consensus_peers.insert(self.cfg.identity, next_epoch);
 
         loop {
-            // a threshold of peers sent verified epochs, so target the next epoch
-            if consensus_peers.len() > self.cfg.max_faulty() * 2 {
+            // if a threshold of peers agree on an epoch, go with that
+            for epoch in consensus_peers.values() {
+                if consensus_peers.values().filter(|e| *e == epoch).count() >= threshold {
+                    return *epoch;
+                }
+            }
+
+            // if all responded return the max of next_epoch
+            if consensus_peers.len() == self.cfg.peers.len() {
+                info!("All peers responded with no consensus epoch");
                 return *consensus_peers.values().max().unwrap();
             }
 
-            match self.connections.receive().await {
-                (peer, EpochMessage::Rejoin(Some(history), epoch)) => {
+            match tokio::time::timeout(timeout, self.connections.receive()).await {
+                Ok((peer, EpochMessage::Rejoin(Some(history), epoch))) => {
                     let is_recent = epoch <= history.outcome.epoch + max_age;
                     if history.verify_sig(&pks).is_ok() && is_recent {
                         consensus_peers.insert(peer, epoch);
                     }
                 }
-                (peer, EpochMessage::Rejoin(None, epoch)) => {
+                Ok((peer, EpochMessage::Rejoin(None, epoch))) => {
                     if epoch <= max_age {
                         consensus_peers.insert(peer, epoch);
                     }
                 }
-                msg => msg_buffer.push(msg),
+                Ok((peer, EpochMessage::RejoinRequest)) => {
+                    let target = Target::Nodes(BTreeSet::from([peer]));
+                    let msg = EpochMessage::Rejoin(self.last_signed_epoch(next_epoch), next_epoch);
+                    self.connections.send(target, msg).await;
+                }
+                Ok(msg) => msg_buffer.push(msg),
+                // if peers had an opportunity to reply take max(next_epoch) from a peer threshold
+                Err(_) => {
+                    if consensus_peers.len() >= threshold {
+                        warn!("Timed-out waiting for all peers, going with max threshold");
+                        return *consensus_peers.values().max().unwrap();
```

Comment:
We shouldn't have unwraps and prefer expect with a short note why it's ok to do.